Test/DafnyTests/TestAttribute.dfy

```diff
@@ -67,3 +67,171 @@ method {:test} PassingTestUsingNoLHSAssignOrHalt() {
 method {:test} FailingTestUsingNoLHSAssignOrHalt() {
   :- expect FailUnless(false);
 }
+
+class Even {
+    var value:int;
+    function method IsValid():bool reads this {
+        this.value % 2 == 0
+    }
+    constructor (value:int) 
+        requires value % 2 == 0
+        ensures this.IsValid()
+    {
+        this.value := value;
+    }
+    
+    function method Next():Even? {
+        null // not implemented yet but can be mocked
+    }
+    
+    function method Sum(a:int, b:int):int {
+        a + b
+    }
+}
+
+method {:extern} {:fresh} freshEven() returns (e:Even) ensures fresh(e)
+
+method {:test} PassingTestUsingFresh() {
+    var e:Even := freshEven();
+    e.value := 4;
+    expect(e.IsValid());
+}
+
+method {:test} FailingTestUsingFresh() {
+    var e:Even := freshEven();
+    e.value := 5;
+    expect(e.IsValid());
+}
+
+method {:extern} {:mock} MockValidEven() returns (e:Even) 
+    ensures fresh(e) 
+    ensures e.IsValid() == true
+method {:extern} {:mock} MockInValidEven() returns (e:Even) 
+    ensures fresh(e) 
+    ensures e.IsValid() == false
+
+method {:test} PassingTestUsingValidMock() {
+    var e:Even := MockValidEven();
+    expect(e.IsValid());
+}
+
+method {:test} PassingTestUsingInValidMock() {
+    var e:Even := MockInValidEven();
+    expect(!e.IsValid());
+}
+
+method {:extern} {:mock} MockSum() returns (e:Even) 
+    ensures fresh(e) 
+    ensures e.Sum(2, 2) == 3
+
+method {:test} PassingTestMockSum() {
+    var e:Even := MockSum();
+    expect(e.Sum(2, 2) == 3);
+}
+
+method {:extern} {:mock} MockSumForall() returns (e:Even) 
+    ensures fresh(e) 
+    ensures forall a:int, b:int :: e.Sum(a, b) == 3
+
+method {:test} PassingTestMockForall() {
+    var e:Even := MockSumForall();
+    expect(e.Sum(2, 2) == 3);
+    expect(e.Sum(3, 2) == 3);
+}
+
+method {:extern} {:mock} MockSumAsMultiplication() returns (e:Even) 
+    ensures fresh(e) 
+    ensures forall a:int :: e.Sum(3, a) == a * 3
+    
+method {:test} PassingTestMockSumAsMultiplication() {
+    var e:Even := MockSumAsMultiplication();
+    expect(e.Sum(3, 2) == 6);
+    expect(e.Sum(3, 0) == 0);
+}
+
+method {:extern} {:mock} MockSumWithArgumentMatcher() returns (e:Even) 
+    ensures fresh(e) 
+    ensures forall a:int, b:int :: (b < a) ==> (e.Sum(a, b) == a * b)
+    ensures forall a:int, b:int :: (b >= a) ==> (e.Sum(a, b) == -a * b)
```

Comment:
I'm not sure forcing the mock to obey the original post-conditions is consistent with what we want the mocks to be able to do. 

Mocks are _not_ intended to be implementations of the original methods; rather, they provide some limited behavior, usually inconsistent with the original method behavior, to achieve some testing goal.

Reply:
Might we include a switch to enable or disable the addition of the original contract for the mocked method?

Reply:
Thanks for the helpful discussion on this yesterday! I still believe that using a mock to drive code to behave in a way that is impossible otherwise is questionable (such as reaching a statement that is otherwise unreachable), but I can appreciate that there is at least some use for this for debugging.

Therefore I am adequately mollified if we add a switch for including the original contract, AND we make it on by default, with a warning in the documentation that turning it off can lead to unsound results.

Reply:
As per the discussion on Slack, I have made it so that Dafny spits an error message if a stubbed method includes any post-postconditions (or if the code tries to mock an object's field) - see the updated test files for an example.